### PR TITLE
Fix python function coercion in FunctionXD special methods.

### PR DESCRIPTION
**Pull request by CnlPepper**
Implements functionality required to solve #325.

### CODE REVIEW

File: raysect/core/math/function/function1d/base.pyx [hunk 1]
```diff
@@ -62,82 +62,82 @@ cdef class Function1D:
         return self.evaluate(x)
 
     def __add__(object a, object b):
-        if isinstance(a, Function1D):
-            if isinstance(b, Function1D):
+        if is_callable(a):
+            if is_callable(b) or callable(b):
```
Comment (jacklovell) [inline raysect/core/math/function/function1d/base.pyx]: Isn't the `or callable (b)` here redundant? This check is already performed in `is_callable(b)`.

File: raysect/core/math/function/function1d/base.pyx [hunk 2]
```diff
@@ -62,82 +62,82 @@ cdef class Function1D:
...
                 # a() + b()
-                return AddFunction1D(<Function1D> a, <Function1D> b)
+                return AddFunction1D(a, b)
             elif isinstance(b, numbers.Real):
                 # a() + B -> B + a()
-                return AddScalar1D(<double> b, <Function1D> a)
+                return AddScalar1D(<double> b, a)
         elif isinstance(a, numbers.Real):
-            if isinstance(b, Function1D):
+            if is_callable(b) or callable(b):
```
Comment (jacklovell) [inline raysect/core/math/function/function1d/base.pyx]: Same here, `is_callable(b)` should be sufficient.

File: raysect/core/math/function/function1d/base.pxd [hunk 3]
```diff
@@ -95,3 +96,13 @@ cdef class PowScalarFunction1D(Function1D):
...
+
+
+cdef inline bint is_callable(object f):
+    """
+    Tests if an object is a python callable or function object.
+    
+    :param object f: Object to test.  
+    :return: True if callable, False otherwise.
+    """
+    return isinstance(f, Function1D) or callable(f)
```
Comment (jacklovell) [inline raysect/core/math/function/function1d/base.pxd]: Is there are set rule for newlines at the end of files? Other files in the project seem to have them, and my editor adds them automatically.

File: raysect/core/math/function/function2d/base.pxd [hunk 4]
```diff
@@ -96,3 +96,13 @@ cdef class PowScalarFunction2D(Function2D):
...
+
+
+cdef inline bint is_callable(object f):
+    """
+    Tests if an object is a python callable or function object.
+    
+    :param object f: Object to test.  
+    :return: True if callable, False otherwise.
+    """
+    return isinstance(f, Function2D) or callable(f)
```
Comment (jacklovell) [inline raysect/core/math/function/function2d/base.pxd]: Same comment about newline at end of file

File: raysect/core/math/function/function3d/base.pxd [hunk 5]
```diff
@@ -95,3 +95,13 @@ cdef class PowScalarFunction3D(Function3D):
...
+
+
+cdef inline bint is_callable(object f):
+    """
+    Tests if an object is a python callable or function object.
+    
+    :param object f: Object to test.  
+    :return: True if callable, False otherwise.
+    """
+    return isinstance(f, Function3D) or callable(f)
```
Comment (jacklovell) [inline raysect/core/math/function/function3d/base.pxd]: Same comment about newline at end of file.

File: raysect/core/math/function/function1d/base.pyx (hunk 1 above)
Comment (CnlPepper) [reply raysect/core/math/function/function1d/base.pyx]: Thanks, that was a find-replace mistake.

**Pull request merged**